### ISSUE TITLE
RobotsTxtMiddleware doesn't support wildcards in Disallow rules

### ISSUE THREAD

**Issue by mattfullerton** [scrapy/scrapy#754]
This is because, as I understand, Python's robotparser module doesn't support them either. There is an alternative, drop-in module, Robotexclusionrulesparser:

http://nikitathespider.com/python/rerp/

This line would need to be changed:
https://github.com/scrapy/scrapy/blob/master/scrapy/contrib/downloadermiddleware/robotstxt.py#L7

**Comment by redapple** [comment]
I believe this is the one: https://github.com/scrapy/scrapy/pull/949
@mattfullerton , are you still working on it?

**Comment by mattfullerton** [comment]
No :blush:

**Comment by Parth-Vader** [comment]
Hi, I'm trying to make the PR for this issue. 
I have made the required changes, and after running the local test `tests/test_downloadermiddleware_robotstxt.py`, one test is failed.

The error is : 

```
twisted.internet.defer.FirstError: FirstError[#1, [Failure instance: Traceback: <class 'twisted.trial.unittest.FailTest'>: did not catch an error, instead got None
/home/parth/Codes/scrapy/.tox/py35/lib/python3.5/site-packages/twisted/internet/defer.py:653:_runCallbacks
/home/parth/Codes/scrapy/scrapy/downloadermiddlewares/robotstxt.py:77:cb
/home/parth/Codes/scrapy/.tox/py35/lib/python3.5/site-packages/twisted/internet/defer.py:459:callback
/home/parth/Codes/scrapy/.tox/py35/lib/python3.5/site-packages/twisted/internet/defer.py:567:_startRunCallbacks
--- <exception caught here> ---
/home/parth/Codes/scrapy/.tox/py35/lib/python3.5/site-packages/twisted/internet/defer.py:653:_runCallbacks
/home/parth/Codes/scrapy/.tox/py35/lib/python3.5/site-packages/twisted/trial/_asynctest.py:67:_cb
]]
```

**Comment by redapple** [comment]
@Parth-Vader , which tests fails?
One can usually see the failing test when running `tox -- -v`:

```
$ tox -- -v tests/test_downloadermiddleware_robotstxt.py
GLOB sdist-make: /home/paul/src/scrapy/setup.py
py27 inst-nodeps: /home/paul/src/scrapy/.tox/dist/Scrapy-1.3.2.zip
py27 installed: appdirs==1.4.3,asn1crypto==0.21.1,attrs==16.3.0,Automat==0.5.0,backports.shutil-get-terminal-size==1.0.0,blessings==1.6,botocore==1.5.24,bpython==0.16,brotlipy==0.6.0,cffi==1.9.1,click==6.7,constantly==15.1.0,coverage==4.3.4,cryptography==1.8.1,cssselect==1.0.1,curtsies==0.2.11,decorator==4.0.11,docutils==0.13.1,enum34==1.1.6,Flask==0.12,funcsigs==1.0.2,greenlet==0.4.12,idna==2.5,incremental==16.10.1,ipaddress==1.0.18,ipython==5.3.0,ipython-genutils==0.1.0,itsdangerous==0.24,Jinja2==2.9.5,jmespath==0.9.2,leveldb==0.194,lxml==3.7.3,MarkupSafe==1.0,mitmproxy==0.10.1,mock==2.0.0,netlib==0.10.1,olefile==0.44,packaging==16.8,parsel==1.1.0,pathlib2==2.2.1,pbr==2.0.0,pexpect==4.2.1,pickleshare==0.7.4,Pillow==4.0.0,pkg-resources==0.0.0,prompt-toolkit==1.0.13,ptyprocess==0.5.1,py==1.4.32,pyasn1==0.2.3,pyasn1-modules==0.0.8,pycparser==2.17,PyDispatcher==2.0.5,Pygments==2.2.0,pyOpenSSL==16.2.0,pyparsing==2.2.0,pytest==2.9.2,pytest-cov==2.2.1,pytest-twisted==1.5,python-dateutil==2.6.0,queuelib==1.4.2,requests==2.13.0,scandir==1.5,Scrapy==1.3.2,service-identity==16.0.0,simplegeneric==0.8.1,six==1.10.0,testfixtures==4.13.5,traitlets==4.3.2,Twisted==17.1.0,urwid==1.3.1,w3lib==1.17.0,wcwidth==0.1.7,Werkzeug==0.12,zope.interface==4.3.3
py27 runtests: PYTHONHASHSEED='1336577254'
py27 runtests: commands[0] | py.test --cov=scrapy --cov-report= -v tests/test_downloadermiddleware_robotstxt.py
============================================================================================================================= test session starts =============================================================================================================================
platform linux2 -- Python 2.7.12+, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /home/paul/src/scrapy/.tox/py27/bin/python2.7
cachedir: .cache
rootdir: /home/paul/src/scrapy, inifile: pytest.ini
plugins: twisted-1.5, cov-2.2.1
collected 9 items 

tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_ignore_robotstxt_request PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_empty_response PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_error PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_garbage PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_immediate_error PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_meta PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_ready_parser PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_settings PASSED

========================================================================================================================== 9 passed in 1.58 seconds ===========================================================================================================================
___________________________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________________________
  py27: commands succeeded
  congratulations :)
```

**Comment by Parth-Vader** [comment]
@redapple 
The second test is failing : 
```
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_ignore_robotstxt_request PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt FAILED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_empty_response PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_error PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_garbage PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_immediate_error PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_meta PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_ready_parser PASSED
tests/test_downloadermiddleware_robotstxt.py::RobotsTxtMiddlewareTest::test_robotstxt_settings PASSED
```

I will look into it.

**Comment by Gallaecio** [comment]
Fixed by #3796. It is now possible to select a different robots.txt parser that does support wildcards in disallow rules, or even write your own parser or your own adapter for an existing parser.
